core/src/main/java/org/elasticsearch/transport/TcpTransport.java

```diff
@@ -1053,6 +1053,23 @@ private BytesReference buildMessage(long requestId, byte status, Version nodeVer
         return new CompositeBytesReference(header, messageBody, zeroCopyBuffer);
     }
 
+    protected abstract int length(Buffer buffer);
```

Comment:
adding another generic here is a big move, I think it's not work it, it's not a justification for not having to create a lightweight BytesReference instance. Yet, we can for instance change the way `validateMessageHeader` works and fix it to look like `validateMessageHeader(StreamInput input, int length)` that way we can just read the bytes form the stream we need instead? not sure it would buy us anything.